docs/architecture/adr-039-epoched-staking.md

```diff
@@ -0,0 +1,110 @@
+
+This can be handled by making a parameter for the epoch pipeline. This parameter should not be alterable except during hard forks, to mitigate implementation complexity of switching the pipeline length.
+
+### Rewards
+
+Even though all staking updates are applied at epoch boundaries, rewards can still be distributed immediately when they are claimed. This is because they do not affect the current stake weights, as we do not implement auto-bonding of rewards. If such a feature were to be implemented, it would have to be setup so that rewards are auto-bonded at the epoch boundary.
+
+## Decision
+
+__Step-1__:  Implement buffering of all staking and slashing messages. 
+
+First we create a pool for storing tokens that are being bonded, but should be applied at the epoch boundary called the `EpochDelegationPool`. Then, we have two separate queues, one for staking, one for slashing. We describe what happens on each message being delivered below:
+
+### Staking messages
+- **MsgCreateValidator**: Move user's self-bond to `EpochDelegationPool` immediately. Queue a message for the epoch boundary to handle the self-bond, taking the funds from the `EpochDelegationPool`. If Epoch execution fail, return back funds from `EpochDelegationPool` to user's account.
+- **MsgEditValidator**: Validate message and if valid queue the message for execution at the end of the Epoch.
+- **MsgDelegate**: Move user's funds to `EpochDelegationPool` immediately. Queue a message for the epoch boundary to handle the delegation, taking the funds from the `EpochDelegationPool`. If Epoch execution fail, return back funds from `EpochDelegationPool` to user's account.
+- **MsgBeginRedelegate**: Validate message and if valid queue the message for execution at the end of the Epoch.
+- **MsgUndelegate**: Validate message and if valid queue the message for execution at the end of the Epoch.
+
+### Slashing messages
+- **MsgUnjail**: Validate message and if valid queue the message for execution at the end of the Epoch.
+- **Slash Event**: Whenever a slash event is created, it gets queued in the slashing module to apply at the end of the epoch. The queues should be setup such that this slash applies immediately.
+
+### Evidence Messages
+ - **MsgSubmitEvidence**: This gets executed immediately, and the validator gets jailed immediately. However in slashing, the actual slash event gets queued.
+
+Then we add methods to the end blockers, to ensure that at the epoch boundary the queues are cleared and delegation updates are applied.
+
+
+__Step-2__: Implement querying of queued staking txs. 
+
+When querying the staking activity of a given address, the status should return not only the amount of tokens staked, but also if there are any queued stake events for that address. This will require nodes supporting querying to either do some more indexing to have this be efficiently queryable, or to have transactions 
+
+__Step-3__: Adjust gas
```

Comment:
Agreed, this is an issue that goes beyond just Epoched Staking. Perhaps there should be a separate discussion/ADR tackling this issue.